### ISSUE TITLE
Filters should be separated

### ISSUE THREAD

**Issue by 9il** [libmir/dcv#85]
See https://en.wikipedia.org/wiki/Separable_filter

**Comment by 9il** [comment]
The idea of separable filter is to perform one direction and then another one.
We also need to add image cache-friendly blocking. Image should be splitted into blocks 15 KBs each. Then for each block we apply H and G filter parts.
In addition we may remove `Yes.vectorized` and `Yes.fastmath` (needs benchmark).

**Comment by 9il** [comment]
see also #86

**Comment by ljubobratovicrelja** [comment]
> We also need to add image cache-friendly blocking. Image should be splitted into blocks 15 KBs each

This is something I've been meaning to talk to you about for some time now. Halide language may come as nice reference, for cache friendliness:
http://halide-lang.org/

Awesome let's do it. Do you have a proposition how do we do this? I could come up with something, but I'd really like go hear what you think. Do we need (or want) to redesign the base filtering API?

**Comment by 9il** [comment]
> Halide language may come as nice reference, for cache friendliness:
> http://halide-lang.org/

Awesome! We _can_ implement Halide as library solution in Dlang :-)

**Comment by ljubobratovicrelja** [comment]
> We can implement Halide as library solution in Dlang

What do you mean by this? Using halide inside D?

**Comment by 9il** [comment]
> What do you mean by this? Using halide inside D?

No, we can implement its functionality as an ndslice module

**Comment by ljubobratovicrelja** [comment]
> we can implement its functionality as ndslice module

YES - This is something I've been thinking about for a long time - we should definitely do this!
